radis/lbl/factory.py

```diff
@@ -1258,6 +1258,11 @@ def non_eq_spectrum(
         # Check line database and parameters, reset populations and scaled line dataframe
         # ----------
         self._check_line_databank()
+        # Checks and loads Energy level database
+        if self.misc.load_energies == False:
+            self._init_rovibrational_energies(self.levels, self.params.levelsfmt)
+            self.misc.load_energies = True
+
```

Comment:
- If I remember well, there is an (underused) non_eq_bands() function in bands.py that would require a similar treatment (it does the same thing as non_eq_spectrum, but only computes some spectral bands, before combining them). 
- same for the `SpectrumFactory.optically_thin_power()` method (also underused : instead of computing a full spectrum it only gives the integral assuming some hypothesis that make the calculation 1000x faster, i.e., that you don't need to resolve every single line) : it also has non-eq options and therefore will fail (the fact it didn't shows there aren't tests there, that's another issue ! :) )

Instead of copying the code here there, what about moving these lines in `self._calc_noneq_parameters()` , which is probably called by all these functions I mentioned ?

Reply:
Actually the test for `SpectrumFactory.optically_thin_power()`  exists in `test_factory.py`  in `test_power_integral()`. I was actually getting error, so I loaded the energy level in that particular test -
sf.load_databank("HITRAN-CO-TEST", db_use_cached=True, `load_energies=True`).
I added the code in `self._calc_noneq_parameters()` and it's working with the default setting. It's also working for `non_eq_bands()`  :+1: 
But there is one case we need to consider, at some places we are accessing the parameters like  `self.parsum_calc`, at some places/tests  like `parfunc = sf.parsum_calc["CO2"][iso]["X"]`  and also some places like [link](https://github.com/radis/radis/blob/c6f544bc07b584389871d7b1a39469cdeb47e537/radis/test/lbl/test_overp.py#L194) without executing `self._calc_noneq_parameters()`. In these cases, we will get errors. So instead of displaying the error, should we automatically run `_init_rovibrational_energies`  based on `load_energies`?

Reply:
In your latest link,  self.get_partition_function_calculator is called anyway, so it will trigger an appropriate error, quite clear for the end-user !

Reply:
Actually, I meant that instead of raising the error we can load the energy levels instead using in the except statement by running `_init_rovibrational_energies()`, if we still encounter an error we can raise it.
The error raised by `get_partition_function_calculator()` will work but there are some cases where we are directly accessing like [link](https://github.com/radis/radis/blob/c6f544bc07b584389871d7b1a39469cdeb47e537/radis/test/levels/test_partfunc.py#L708)
Here also we will get a `Key Error` but this won't be raised by `get_partition_function_calculator()`. I solved it by loading the energy level database. But how will we trigger an appropriate error for this case?

Reply:
Hello ! In the above link `sf.parsum_calc["CO2"][1]["X"].df` is only accessed for test purpose, in a way that's used only by people knowledgable of the code structure. Therefore it's ok if the error is not very clear. You should just enforce load_energies=True in the test itself.